bundles/org.openhab.binding.mqtt.espmilighthub/src/main/java/org/openhab/binding/mqtt/espmilighthub/internal/handler/EspMilightHubHandler.java

```diff
@@ -168,6 +220,117 @@ private void processIncomingState(String messageJSON) {
+        }
+        return result;
+    }
+
+    // https://www.jkps.or.kr/journal/download_pdf.php?spage=865&volume=41&number=6 (8) and (9)
+    private static HSBType calculateHSBFromColorTemp(int mireds, PercentType brightness) {
+        var cct = BIG_DECIMAL_MILLION.divide(new BigDecimal(mireds), MathContext.DECIMAL128);
+        var cctInt = cct.intValue();
+
+        BigDecimal[] coefficients;
+        // 1667K to 4000K and 4000K to 25000K; no range checks since our mired range fits within this
+        if (cctInt <= 4000) {
+            coefficients = KANG_X_COEFFICIENTS[1];
+        } else {
+            coefficients = KANG_X_COEFFICIENTS[0];
+        }
+        var x = polynomial_fit(BIG_DECIMAL_THOUSAND.divide(cct, MathContext.DECIMAL128), coefficients);
+
+        if (cctInt <= 2222) {
+            coefficients = KANG_Y_COEFFICIENTS[2];
+        } else if (cctInt <= 4000) {
+            coefficients = KANG_Y_COEFFICIENTS[1];
+        } else {
+            coefficients = KANG_Y_COEFFICIENTS[0];
+        }
+        var y = polynomial_fit(x, coefficients);
+        var rawHsb = HSBType.fromXY(x.floatValue() * 100.0f, y.floatValue() * 100.0f);
+        return new HSBType(rawHsb.getHue(), rawHsb.getSaturation(), brightness);
+    }
+
+    // https://www.waveformlighting.com/tech/calculate-color-temperature-cct-from-cie-1931-xy-coordinates/
+    private static int calculateColorTempFromHSB(HSBType hsb) {
+        var xy = hsb.toXY();
+        var x = xy[0].toBigDecimal().divide(BIG_DECIMAL_100);
+        var y = xy[1].toBigDecimal().divide(BIG_DECIMAL_100);
+        var n = x.subtract(BIG_DECIMAL_03320).divide(BIG_DECIMAL_01858.subtract(y), MathContext.DECIMAL128);
+        var CCT_K = polynomial_fit(n, MCCAMY_COEFFICIENTS);
+        return BIG_DECIMAL_MILLION.divide(CCT_K, MathContext.DECIMAL128).round(new MathContext(0)).intValue();
+    }
+
+    // https://cormusa.org/wp-content/uploads/2018/04/CORM_2011_Calculation_of_CCT_and_Duv_and_Practical_Conversion_Formulae.pdf
+    // page 19
+    private static BigDecimal calculateDuvFromHSB(HSBType hsb) {
+        var xy = hsb.toXY();
+        var x = xy[0].toBigDecimal().divide(BIG_DECIMAL_100);
+        var y = xy[1].toBigDecimal().divide(BIG_DECIMAL_100);
+        var u = BIG_DECIMAL_4.multiply(x).divide(
+                BIG_DECIMAL_2.multiply(x).negate().add(BIG_DECIMAL_12.multiply(y).add(BIG_DECIMAL_3)),
+                MathContext.DECIMAL128);
+        var v = BIG_DECIMAL_6.multiply(y).divide(
+                BIG_DECIMAL_2.multiply(x).negate().add(BIG_DECIMAL_12.multiply(y).add(BIG_DECIMAL_3)),
+                MathContext.DECIMAL128);
+        var Lfp = u.subtract(BIG_DECIMAL_0292).pow(2).add(v.subtract(BIG_DECIMAL_024).pow(2))
+                .sqrt(MathContext.DECIMAL128);
+        var a = new BigDecimal(
+                Math.acos(u.subtract(BIG_DECIMAL_0292).divide(Lfp, MathContext.DECIMAL128).doubleValue()));
+        var Lbb = polynomial_fit(a, CORM_COEFFICIENTS);
```

Comment:
Same here

Reply:
I've changed from `var` to `BigDecimal` here to make the type apparent, but I'm leaving the name capitalized since it matches precisely with the reference formulas.